### ISSUE TITLE
IntelliSense: Drill in on callback?

### ISSUE THREAD

**Issue by egamma** [Microsoft/TypeScript#8134]
_From @unional on April 14, 2016 23:32_

![image](https://cloud.githubusercontent.com/assets/3254987/14546974/4f88973e-025e-11e6-8053-724cc8ecc51e.png)
I would be great if when `tape((<cursor>))` it would be able to drill in the `tape.TestCase` and show the signature of the callback. :tulip: 

This is probably on `tsc`, but let's start here. :smile:

_Copied from original issue: Microsoft/vscode#5284_

**Comment by egamma** [comment]
Yes, hint is computed by Salsa/TypeScript. Moving to TypeScript.

**Comment by mhegazy** [comment]
We will need to see if the current expression is a parenthesized expression and should the signature help at the first parameter position instead of the main call expression.

**Comment by omefire** [comment]
I'm having issues reproing this bug.
@unional, Could you please let me know what the definition of tape.TestCase is ?

**Comment by unional** [comment]
For example:
![image](https://cloud.githubusercontent.com/assets/3254987/17381525/37d7496c-5981-11e6-97f6-9d7446cdafb8.png)
```ts

interface Callback {
  (a: string, b: string): void;
}

function foo(cb: Callback) {
}

foo(())
```

**Comment by omefire** [comment]
Thanks! The example provided is indeed helpful.

**Comment by DanielRosenwasser** [comment]
Looks like you'd only do this if you have a single-call signature. Otherwise it doesn't really make a ton of sense to do so.

**Comment by unional** [comment]
It is used quite a lot in typings. For example
https://github.com/types/npm-bluebird/blob/master/bluebird.d.ts#L86
you would be templated to create:
```ts
interface SomeCallback {
  (err: any, value?: R) => void
}
```
so you can reuse them on those functions.

One example:
https://github.com/types/env-node/blob/master/6/node.d.ts#L850

**Comment by DanielRosenwasser** [comment]
Sure, I've had this experience as well, so I definitely understand the advantages. I'm just mentioning that if you have any members other than call signatures, or if you have any overloads, we should not use the shorthand.

**Comment by unional** [comment]
Daniel, maybe I'm not understand what do you mean on the single call signature. 

It does seem that we are on the same page :). 
I just want to understand what do you mean so I may learn some new things :smile:

**Comment by DanielRosenwasser** [comment]
Sure thing @unional. All objects have a set of *members*. These aren't the same things as *properties*. A property is a type of member, but members also include:

* call signatures
* construct signatures
* index signatures

The thing that allows a type to be callable is a *call signature*. You might already be familiar with call signatures if you've thought about overloads.

Here is an example of a type with a single call signature:

```ts
interface Foo {
    (hello: string, world: number): void;
}
```

If I have something of type `Foo`, I can call it with a `string` and `number`.

```ts
let f: Foo = /* ... */;

f("hello!", 100);
```

You can also have *multiple call signatures*.

```ts
interface Foo {
    (hello: string, world: number): void;
    (hello: string): void;
}
```

If `Foo` was declared like that, you could have also called `f` like so:

```ts
f("hello!");
```

_______

We have two different shorter ways of writing things with call signatures. One place is method members.

If you have something like

```ts
interface Bar {
    m(): void;
}
```

Here, `Bar` is actually equivalent to

```ts
interface Bar {
    m: {
        (): void;
    }
}
```

Meaning that `Bar` has a property named `m`, which has a single call signature.

Another short-hand that we use is arrow function types. These are just objects with a single call signature.

In other words, if you've ever seen `(x: number) => void`, that's actually just an easier way of writing the type `{ (x: number): void }`.

Which is why I think that if you have an interface type with no members apart from a single call signature, we should consider showing the arrow function notation.

_______

For more about call signatures, you can check out the language spec too: https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3.9.2

**Comment by unional** [comment]
@DanielRosenwasser many thanks! Great examples. Especially that `Bar` and the arrow function one. Didn't know function signature is expanded to a property with single call signature (in this case). It does make a lot of sense.
